test/extended/util/framework.go

```diff
@@ -1310,6 +1311,52 @@ func (r *podExecutor) Exec(script string) (string, error) {
 	return out, waitErr
 }
 
+type GitRepo struct {
+	upstream     git.Repository
+	upstreamPath string
+	repo         git.Repository
+	repoPath     string
+}
+
+//TODO: add comment
+func (r GitRepo) AddAndCommit(file, content string) error {
+	if err := ioutil.WriteFile(filepath.Join(r.repoPath, file), []byte(content), 0666); err != nil {
+		return err
+	}
+	if err := r.repo.Add(r.repoPath, file); err != nil {
+		return err
+	}
+	if err := r.repo.Commit(r.repoPath, "added file "+file); err != nil {
+		return err
+	}
+	return nil
+}
+
+//TODO: add comment
+func (r GitRepo) Cleanup() {
+	if r.upstreamPath != "" {
+		os.RemoveAll(r.upstreamPath)
+	}
+	if r.repoPath != "" {
+		os.RemoveAll(r.repoPath)
+	}
+}
+
+//TODO: add comment
+func NewGitRepo(repoPath string) (GitRepo, error) {
```

Comment:
Unfortunately, I think we need upstream repo as well and it was required from `oc new-app` itself.
I found two ways providing source in git repo:
1) URL - and I would like to try to avoid this due to reasons from https://github.com/openshift/origin/pull/19332#discussion_r181662278
2) local path - in this case, it needs to be git repo with upstream (I will try to find the code or at least link to docs, I think I saw it there as well)

Reply:
I think that, using local path, the repo has to have an accessible origin, not an upstream.

Reply:
and I am thinking that if it doesn't it will tar up the contents and upload it?  @bparees might know better.

Reply:
it needs to have a valid remote. i'm not sure what rules we apply beyond that to pick *which* remote if there are several available.

if you don't have a valid remote, you get a binary build, which in this case might actually be fine.

Reply:
again, if we simplify this test to not actually run the build, this probably gets easier.

Reply:
yeah i guess that makes sense since you can't do a binary build w/ a pipeline buildconfig.